mm2src/common/custom_futures/repeatable.rs

```diff
@@ -0,0 +1,785 @@
+                will_be_after_timeout < timestamp_ms
+            },
+        }
+    }
+}
+
+impl<Factory, F: Unpin, T, E> Unpin for Repeatable<Factory, F, T, E> {}
+
+impl<Factory, F, T, E> Future for Repeatable<Factory, F, T, E>
+where
+    Factory: FactoryTrait<F>,
+    F: RepeatableTrait<T, E>,
+{
+    type Output = Result<T, RepeatError<E>>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        loop {
+            if poll_timeout(&mut self.timeout_fut, cx).is_pending() {
+                return Poll::Pending;
+            }
+
+            let mut exec = self.exec_fut.take().unwrap();
+            match exec.poll_unpin(cx) {
+                Poll::Ready(Action::Ready(ready)) => return Poll::Ready(Ok(ready)),
+                Poll::Ready(Action::Retry(error)) => {
+                    if let Some(ref inspect) = self.inspect_err {
+                        inspect(&error);
+                    }
+
+                    if self.until.is_none() && self.attempts.is_none() {
+                        // We should try to execute the future only once
+                        // if neither `self.until` nor `self.attempts` are specified.
+                        // https://github.com/KomodoPlatform/atomicDEX-API/pull/1564#discussion_r1050778208
+                        return Poll::Ready(Err(RepeatError::attempts(1, error)));
+                    }
+
+                    if let Some(until) = self.until {
+                        if !self.check_can_retry_after_timeout(until) {
+                            return Poll::Ready(Err(RepeatError::timeout(until, error)));
+                        }
+                    }
+
+                    if let Some(ref mut attempts) = self.attempts {
+                        // Check if we have one more attempt to retry to execute the future.
+                        attempts.current_attempt += 1;
+                        if attempts.current_attempt >= attempts.total_attempts {
+                            return Poll::Ready(Err(RepeatError::attempts(attempts.current_attempt, error)));
+                        }
+                    }
+
+                    // Create a new future attempt.
+                    self.exec_fut = Some((self.factory)());
+                    // Set the timeout future if `self.repeat_every` is specified,
+                    // otherwise the `self.exec_fut` will be repeated right at the next iteration.
+                    self.timeout_fut = self.repeat_every.map(|timeout| Timer::sleep(timeout.as_secs_f64()));
+                    continue;
+                },
+                // We should proceed with this `exec` future attempt.
+                Poll::Pending => {
+                    self.exec_fut = Some(exec);
+                    return Poll::Pending;
+                },
+            }
+        }
+    }
+}
+
+struct AttemptsState {
+    current_attempt: usize,
+    total_attempts: usize,
```

Comment:
I suppose it should be `NonZeroUsize`.

Reply:
Same comment as this https://github.com/KomodoPlatform/atomicDEX-API/pull/1564#discussion_r1056415366